whats-that-typeclass-functor/article.md

```diff
@@ -0,0 +1,467 @@
+
+```haskell
+> :info (->)
+type (->) :: * -> * -> *
+data (->) a b
+```
+
+`(->)` is parametrized with two types — `(->) a b`. In other words, it is `a -> b` — a function with one argument.
+
+We know that Functor can be implemented for types with the  `* -> *` kind. Unfortunately, the kind of `(->)` is `* -> * -> *`. It doesn’t satisfy the Functor instance in this form because a Functor can only have one type argument.
+
+Hence, it’s required to apply it once, like `Either a` and `((,) a)` do. In the `(->)` case, it means providing the function argument’s type — `(->) a` or `a ->` (the syntax of the latter is not allowed in Haskell, though).
+
+For example, functions with the following type signatures have correct Functor instances:
+
+```haskell
+Int -> a
+String -> a
+(Char, Int) -> a
+[Int] -> a
+```
+
+We found out what functions could be functors and what data they match. So, the next question — what does it mean to `fmap` the function? Intuitively, it’s about changing the function, i.e., the action it performs. But what is the fixed context here?
+
+Let’s construct the type definition of `fmap` for a one-argument function. For an arbitrary Functor `f`, `fmap` is:
+
+```haskell
+fmap :: Functor f => (a -> b) -> f a -> f b
+```
+
+The type of our data is `(->) c`. Therefore, we need to replace `f t` by `(->) c t` or `c -> t`. Consequently, the type signature is:
+
+```haskell
+fmap :: (a -> b) -> (c -> a) -> (c -> b)
+```
+
+Now we can see that the fixed context corresponds to the type of function’s argument `c`. And the value being modified is the function’s return type — we go from `a` to `b` with the help of the `a -> b` function.
+
+All in all, `fmap` for functions transforms what the function returns while keeping the input unchanged! Does that look familiar to you?
+
+<details>
+<summary>Which Haskell function or operator can produce a <code>c -> b</code> function from <code>a -> b</code> and <code>c -> a</code>?</summary>
+
+Function composition:
+
+```haskell
+> :t (.)
+(.) :: (a -> b) -> (c -> a) -> c -> b
+```
+
+Note that
+
+`(a -> b) -> (c -> a) -> c -> b`
+
+is exactly the same as
+
+`(a -> b) -> (c -> a) -> (c -> b)`
+
+since `->` in Haskell is right-associative.
+
+<hr>
+</details>
+
+Cool! `fmap` for Functor is just the composition of functions.
+
+Here’s the `((->) a)` instance implementation:
+
+```haskell
+instance Functor ((->) a) where
+  fmap = (.)
+```
+
+In conclusion, let’s make sure `fmap` is `(.)` for one-argument function `a -> b` with the example below:
+
+```haskell
+-- fmap :: (Int -> Int) -> (Int -> Int) -> Int -> Int
+> fmap (+1) (*2) 2
+5
+> (+1) . (*2) $ 2
+5
+```
+
+In fact, it’s uncommon to consider a function a Functor in Haskell, it’s more of a fancy case. However, you’ll not be puzzled when you come across magic and strange usage of `fmap` in the future!
+
+## Why do you need to know Functor?
+
+In general, Functor is not a really powerful typeclass. Its methods are easy to grasp and to implement.
```

Comment:
>Functor is not a really powerful typeclass. Its methods are easy to grasp and to implement.

Not sure about this statement. I don't think the methods being easy to grasp have anything to do with the usefulness of the abstraction. Is the methods in Monad typeclass much more complicated than the Functor methods? Nevertheless, Monads are pretty powerful, Right?

Reply:
Rephrased a bit

Reply:
I think this section "Why do you need to know Functor", can be reduced to something that amounts to "Because it enables one to apply transformation on the wrapped type without knowing anything about the wrapper, and it is a very useful thing".

Reply:
[Added](https://github.com/serokell/blog-posts/commit/f47a71dbca41039eed9cdc3acfa33cb233318e82#diff-4d5bf4bf274bf0b54f99958d24e2d64f6cefc8743f5c0c21150e83148c2e000dR369)